// frappe-js-sdk/lib/auth/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrappeAuth = void 0;
var FrappeAuth = /** @class */ (function () {
    function FrappeAuth(appURL, axios, useToken, token, tokenType) {
        this.appURL = appURL;
        this.axios = axios;
        this.useToken = useToken !== null && useToken !== void 0 ? useToken : false;
        this.token = token;
        this.tokenType = tokenType;
    }
    /** Logs in the user using username and password */
    FrappeAuth.prototype.loginWithUsernamePassword = function (credentials) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.axios
                        .post('/api/method/login', {
                        usr: credentials.username,
                        pwd: credentials.password,
                        otp: credentials.otp,
                        tmp_id: credentials.tmp_id,
                        device: credentials.device,
                    })
                        .then(function (res) { return res.data; })
                        .catch(function (error) {
                        var _a, _b;
                        throw __assign(__assign({}, error.response.data), { httpStatus: error.response.status, httpStatusText: error.response.statusText, message: (_a = error.response.data.message) !== null && _a !== void 0 ? _a : 'There was an error while logging in', exception: (_b = error.response.data.exception) !== null && _b !== void 0 ? _b : '' });
                    })];
            });
        });
    };
    /** Gets the currently logged in user */
    FrappeAuth.prototype.getLoggedInUser = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.axios
                        .get('/api/method/frappe.auth.get_logged_user')
                        .then(function (res) { return res.data.message; })
                        .catch(function (error) {
                        var _a;
                        throw __assign(__assign({}, error.response.data), { httpStatus: error.response.status, httpStatusText: error.response.statusText, message: 'There was an error while fetching the logged in user', exception: (_a = error.response.data.exception) !== null && _a !== void 0 ? _a : '' });
                    })];
            });
        });
    };
    /** Logs the user out */
    FrappeAuth.prototype.logout = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.axios
                        .post('/api/method/logout', {})
                        .then(function () {
                        return;
                    })
                        .catch(function (error) {
                        var _a, _b;
                        throw __assign(__assign({}, error.response.data), { httpStatus: error.response.status, httpStatusText: error.response.statusText, message: (_a = error.response.data.message) !== null && _a !== void 0 ? _a : 'There was an error while logging out', exception: (_b = error.response.data.exception) !== null && _b !== void 0 ? _b : '' });
                    })];
            });
        });
    };
    /** Sends password reset email */
    FrappeAuth.prototype.forgetPassword = function (user) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.axios
                        .post('/', {
                        cmd: 'frappe.core.doctype.user.user.reset_password',
                        user: user,
                    })
                        .then(function () {
                        return;
                    })
                        .catch(function (error) {
                        var _a, _b;
                        throw __assign(__assign({}, error.response.data), { httpStatus: error.response.status, httpStatusText: error.response.statusText, message: (_a = error.response.data.message) !== null && _a !== void 0 ? _a : 'There was an error sending password reset email.', exception: (_b = error.response.data.exception) !== null && _b !== void 0 ? _b : '' });
                    })];
            });
        });
    };
    return FrappeAuth;
}());
exports.FrappeAuth = FrappeAuth;
